### DISCUSSION TITLE
When should I use `clone_weak()`?

### DISCUSSION THREAD

**Discussion by ngdangtu-vn** [bevyengine/bevy#9567] [Q&A]
My problem starts from here:

> https://gitlab.com/ndt-lab/rs-fruits-catcher/-/tree/1fbb9c0c321caad236effec7dfe898d894409433

If I use weak clone on my font handle, it will be easily killed off without a warning. The result is it only display TextSection at first and not do any re-rendering on the later update schedule. ([this problem found by sird0rius](https://www.reddit.com/r/bevy/comments/1606oza/comment/jxlcp6s))

At first, I thought using weak clone with help me optimize the game a bit in loading resource. But it seems like to [be removed early without a probable warning](https://bevy-cheatbook.github.io/assets/handles.html#reference-counting-strong-and-weak-handles), at least from my observation. So I have been wondering if there is any use cases for weak clone, or the function is broken in my case or my font is bad. I don't know which one.

Hope to find an answer in here :D  
Thanks in advance

**Comment by rlidwka** [comment]
> So I have been wondering if there is any use cases for weak clone

Weak clones are used you want to maintain a list of fonts that were loaded previously, regardless of whether they exists now.

If you are actually *using* a font, you have to use strong handle, and you already seem to figure out what happens if you don't.

**Reply by ngdangtu-vn** [reply]
If there is a list of fonts that is not exist. What are benefits from it?

**Reply by rlidwka** [reply]
For example, if I want to maintain the list of all currently used fonts (to display "currently loaded font" list in UI), I would clone each handle into a dedicated resource as a weak handle. Strong handle would not be suitable for this task, because fonts would never get deallocated.

**Reply by ngdangtu-vn** [reply]
I think I still not understand the word "maintain" in here. What are we exactly doing with it? Like we keep it in the list so we can do strong clone it later or anything else?

I'm sorry, I'm totally new to this area and really have no idea how game is making.

**Comment by d-bucur** [comment]
I think that in this case the engine should at least emit an error that it is trying to use an unloaded asset, instead of failing silently.

**Reply by ngdangtu-vn** [reply]
I think so, otherwise it would be confused. Who would have known if we decide to use it later right?
